compiler/rustc_trait_selection/src/solve/trait_goals/structural_traits.rs

```diff
@@ -235,3 +241,114 @@ pub(crate) fn extract_tupled_inputs_and_output_from_callable<'tcx>(
         }
     }
 }
+
+/// Assemble a list of predicates that would be present on a theoretical
+/// user impl for an object type. These predicates must be checked any time
+/// we assemble a built-in object candidate for an object type, since they
+/// are not implied by the well-formedness of the type.
+///
+/// For example, given the following traits:
+///
+/// ```rust,ignore (theoretical code)
+/// trait Foo: Baz {
+///     type Bar: Copy;
+/// }
+///
+/// trait Baz {}
+/// ```
+///
+/// For the dyn type `dyn Foo<Item = Ty>`, we can imagine there being a
+/// pair of theoretical impls:
+///
+/// ```rust,ignore (theoretical code)
+/// impl Foo for dyn Foo<Item = Ty>
+/// where
+///     Self: Baz,
+///     <Self as Foo>::Bar: Copy,
+/// {
+///     type Bar = Ty;
+/// }
+///
+/// impl Bar for dyn Foo<Item = Ty> {}
+/// ```
+///
+/// However, in order to make such impls well-formed, we need to do an
+/// additional step of eagerly folding the associated types in the where
+/// clauses of the impl. In this example, that means replacing
+/// `<Self as Foo>::Bar` with `Ty` in the first impl.
+pub(crate) fn predicates_for_object_candidate<'tcx>(
+    infcx: &InferCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    trait_ref: ty::TraitRef<'tcx>,
+    object_bound: &'tcx ty::List<ty::PolyExistentialPredicate<'tcx>>,
+) -> Vec<ty::Predicate<'tcx>> {
+    let mut requirements = vec![];
+    requirements.extend(
+        infcx
+            .tcx
+            .super_predicates_of(trait_ref.def_id)
+            .instantiate(infcx.tcx, trait_ref.substs)
+            .predicates,
+    );
+    for item in infcx.tcx.associated_items(trait_ref.def_id).in_definition_order() {
+        if item.kind == ty::AssocKind::Type {
+            requirements
+                .extend(infcx.tcx.item_bounds(item.def_id).subst(infcx.tcx, trait_ref.substs));
+        }
+    }
+
+    let mut replace_projection_with = FxHashMap::default();
+    for bound in object_bound {
+        if let ty::ExistentialPredicate::Projection(proj) = bound.skip_binder() {
+            let proj = proj.with_self_ty(infcx.tcx, trait_ref.self_ty());
+            let old_ty = replace_projection_with.insert(proj.def_id(), bound.rebind(proj));
+            assert_eq!(
+                old_ty,
+                None,
+                "{} has two substitutions: {} and {}",
```

Comment:
I think that assert can be triggered :thinking: but I also have to admit that I don't really care :grin: can change it to a `delay_span_bug` when it does get triggered

Reply:
There are some ways that this can be triggered that causes ICEs in the current solver, when you have a supertrait bound like:

```
trait Foo: Bar<Assoc = i32> {}
```

and `&dyn Foo<Assoc = ?0>`, which will create `Assoc = ?0` and `Assoc = i32` as object bounds.

But yeah, I'll leave it as an assert for now.